netket/optimizer/sr/base.py

```diff
@@ -32,3 +34,96 @@ class SR:
 
     diag_shift: float = 0.01
     """Diagonal shift added to the S matrix."""
+
+    def create(self, vstate, **kwargs) -> "AbstractSMatrix":
+        """
+        Construct the Lazy representation of the S corresponding to this SR type.
+
+        Args:
+            vstate: The Variational State
+        """
+        raise NotImplementedError
+
+
+@struct.dataclass
+class AbstractSMatrix:
+    """
+    S matrix base class.
+    This can either be a jnp matrix, a lazy wrapper, or anything, as long as
+    it satisfies this basic API.
+
+    An AbstractSMatrix must support the following API:
+
+    - :code:`__matmul__(y)`, meaning you must be able to do S@vec, where vec is
+        either a PyTree of parameters or it's dense ravelling.
+    - :code:`solve(y, **kwargs)`, which must solve the linear system Sx=y with
+        any method available, usually but not necessarily defined inside the
+        sr object stored inside the AbstractSMatrix.
+        This function must accept arbitrary additional arguments.
+
+
+    Additionally, you get for free:
+    - :code:`__call__(y)`, meaning you must be able to do S(vec) = S@vec.
+        This is defined by the base class so you don't need to define it.
+        This guarantees that you can pass this matrix to sparse solvers.
+    - :code:`__rtruediv__(y)`, that allows you to solve Sx=y by doing y\\S.
```

Comment:
Having `x = y \\ S` solve `Sx = y` is a bit confusing: Julia (and matlab) uses `A\b` to solve `Ax = b`, so exactly the other way round. I think this is more difficult to implement here, so we could just keep `solve`. It's not used so often that a shorthand is really necessary IMHO.

Reply:
This is `__rtruediv__` so happens when you call `b/A`. But indeed you are right that this is a bit too much.